In [1]:
import numpy as np
import adaboost
import utils

# 一个简单的人造数据

In [2]:
datMat,classLabels=utils.loadSimpData()
print ("datMat = ", datMat)
print ("classLabels = ", classLabels)

datMat =  [[1.  2.1]
 [2.  1.1]
 [1.3 1. ]
 [1.  1. ]
 [2.  1. ]]
classLabels =  [1.0, 1.0, -1.0, -1.0, 1.0]


Expect Result:  

datMat =  [[1.  2.1]  
 [2.  1.1]  
 [1.3 1. ]  
 [1.  1. ]  
 [2.  1. ]]  
classLabels =  [1.0, 1.0, -1.0, -1.0, 1.0]

In [3]:
D = np.ones(5)/5   # 平均权重
bestStump, bestClasEst, minError = adaboost.buildStump(datMat,classLabels,D)
print ("bestStump = ", bestStump)
print ("bestClasEst = ", bestClasEst)
print ("minError = ", minError)

bestStump =  {'feature': 0, 'value': 1.3, 'ineq': 'lt'}
bestClasEst =  [-1.  1. -1. -1.  1.]
minError =  0.2


Expect Result:  

bestStump =  {'feature': 0, 'value': 1.3, 'ineq': 'lt'}  
bestClasEst =  [-1.  1. -1. -1.  1.]  
minError =  0.2  

In [4]:
datArr,labelArr = utils.loadSimpData()
weekClassifier, aggPredict = adaboost.adaBoostTrainDS(datMat,classLabels,9)
print ("weekClassifier = ", weekClassifier)
print ("aggPredict = ", aggPredict)

D= [0.2 0.2 0.2 0.2 0.2]
predict:  [-1.  1. -1. -1.  1.]
aggpredict= [-0.69314718  0.69314718 -0.69314718 -0.69314718  0.69314718]
total error:  0.2
D= [0.5   0.125 0.125 0.125 0.125]
predict:  [ 1.  1. -1. -1. -1.]
aggpredict= [ 0.27980789  1.66610226 -1.66610226 -1.66610226 -0.27980789]
total error:  0.2
D= [0.28571429 0.07142857 0.07142857 0.07142857 0.5       ]
predict:  [1. 1. 1. 1. 1.]
aggpredict= [ 1.17568763  2.56198199 -0.77022252 -0.77022252  0.61607184]
total error:  0.0
weekClassifier =  [{'feature': 0, 'value': 1.3, 'ineq': 'lt', 'alpha': 0.6931471805599453}, {'feature': 1, 'value': 1.0, 'ineq': 'lt', 'alpha': 0.9729550745276565}, {'feature': 0, 'value': 0.9, 'ineq': 'lt', 'alpha': 0.8958797346140273}]
aggPredict =  [ 1.17568763  2.56198199 -0.77022252 -0.77022252  0.61607184]


Expect Result:  

D= [0.2 0.2 0.2 0.2 0.2]  
predict:  [-1.  1. -1. -1.  1.]  
aggpredict= [-0.69314718  0.69314718 -0.69314718 -0.69314718  0.69314718]  
total error:  0.2  
D= [0.5   0.125 0.125 0.125 0.125]  
predict:  [ 1.  1. -1. -1. -1.]  
aggpredict= [ 0.27980789  1.66610226 -1.66610226 -1.66610226 -0.27980789]  
total error:  0.2  
D= [0.28571429 0.07142857 0.07142857 0.07142857 0.5       ]  
predict:  [1. 1. 1. 1. 1.]  
aggpredict= [ 1.17568763  2.56198199 -0.77022252 -0.77022252  0.61607184]  
total error:  0.0  
weekClassifier =  [{'feature': 0, 'value': 1.3, 'ineq': 'lt', 'alpha': 0.6931471805599453}, {'feature': 1, 'value': 1.0, 'ineq': 'lt', 'alpha': 0.9729550745276565}, {'feature': 0, 'value': 0.9, 'ineq': 'lt', 'alpha': 0.8958797346140273}]  
aggPredict =  [ 1.17568763  2.56198199 -0.77022252 -0.77022252  0.61607184]  

In [5]:
adaboost.adaClassify([[5, 5],[0,0]],classifierArr)

NameError: name 'classifierArr' is not defined

[ 2.56198199 -2.56198199]  
array([ 1., -1.])

# Horse Colic Data Test

In [ ]:
datArr,labelArr = utils.loadDataSet('horseColicTraining2.txt')
classifierArray, _ = adaboost.adaBoostTrainDS(datArr,labelArr,10)
testArr,testLabelArr = adaboost.loadDataSet('horseColicTest2.txt')
prediction10 = adaboost.adaClassify(testArr,classifierArray)
accuracy = utils.calculateAccuray(prediction10, testLabelArr)
print ("accuracy = ", accuracy)

In [ ]:
datArr,labelArr = utils.loadDataSet('horseColicTraining2.txt')
classifierArray, _ = adaboost.adaBoostTrainDS(datArr,labelArr,10)
testArr,testLabelArr = adaboost.loadDataSet('horseColicTest2.txt')
prediction10 = adaboost.adaClassify(testArr,classifierArray)

In [ ]:
np.sum(prediction10!=np.array(testLabelArr))

16

参考：https://windmising.gitbook.io/liu-yu-bo-play-with-machine-learning/10-1/10-7

In [ ]:
datArr,labelArr = adaboost.loadDataSet('horseColicTraining2.txt')

In [ ]:
classifierArray,aggClassEst = adaboost.adaBoostTrainDS(datArr,labelArr,10)

In [ ]:
def plotROC(predStrengths, classLabels):
    import matplotlib.pyplot as plt
    total = np.array([np.sum(np.array(classLabels) == 1), np.sum(np.array(classLabels) == -1)])
    cur = total.copy()
    rocNode = [cur/total]
    print (rocNode)
    sortedIndices = predStrengths.argsort()
    print (predStrengths)
    print (sortedIndices)
    for index in sortedIndices:
        if classLabels[index] == 1:
            cur[0] -=1
        else:
            cur[1] -=1
        rocNode.append(cur/total)
    print (np.array(rocNode))
    plt.plot(np.array(rocNode)[:,1], np.array(rocNode)[:, 0])
    plt.show()

In [ ]:
plotROC(aggClassEst,labelArr)